<a href="https://colab.research.google.com/github/Lis30/Homeworks/blob/main/%D0%B4%D0%BE%D0%BC%D0%B0%D1%88%D0%BA%D0%B0_2_%D0%9A%D0%B0%D1%80%D0%BF%D0%BE%D0%B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1) Начнем с простого вопроса: что произойдет в результате выполнения следующей команды?<br>
import pandas as pnds

Импортируется библиотека pandas, но с сокращением pnds, вместо принятого pd

In [ ]:
import pandas as pd

2) Теперь поработаем с данными о поездках на такси в Нью-Йорке, где также имеется информация о погодных условиях и выходных днях.<br>  
pickup_dt – период с точностью до часа<br>
pickup_month – месяц<br>
borough – район Нью-Йорка, из которого был сделан заказ (5 районов + аэропорт)<br>
pickups – число поездок за период (час)<br>
hday – является ли день праздничным/выходным; Y - да,  N - нет<br>
spd – скорость ветра в милях в час<br>
vsb – видимость<br>
temp – температура в градусах Фаренгейта<br>
dewp – точка росы по Фаренгейту<br>
slp – давление<br>
pcp_01 – количество осадков за час<br>
pcp_06 – количество осадков за 6 часов<br>
pcp_24 – количество осадков за 24 часа<br>
sd – глубина снега в дюймах<br>
Импортируйте библиотеку pandas, а затем загрузите датасет 2_taxi_nyc (находится в папке). Результат сохраните в переменную taxi.<br>
Для загрузки датасэта в pandas по ссылке просто выполните pd.read_csv('path')<br>

In [ ]:
taxi = pd.read_csv('2_taxi_nyc.csv')
taxi.head(10)

pickup_dt pickup_month        borough  pickups hday  spd   vsb  \
0  2015-01-01 01:00:00          Jan          Bronx      152    Y  5.0  10.0   
1  2015-01-01 01:00:00          Jan       Brooklyn     1519    Y  5.0  10.0   
2  2015-01-01 01:00:00          Jan            EWR        0    Y  5.0  10.0   
3  2015-01-01 01:00:00          Jan      Manhattan     5258    Y  5.0  10.0   
4  2015-01-01 01:00:00          Jan         Queens      405    Y  5.0  10.0   
5  2015-01-01 01:00:00          Jan  Staten Island        6    Y  5.0  10.0   
6  2015-01-01 01:00:00          Jan            NaN        4    Y  5.0  10.0   
7  2015-01-01 02:00:00          Jan          Bronx      120    Y  3.0  10.0   
8  2015-01-01 02:00:00          Jan       Brooklyn     1229    Y  3.0  10.0   
9  2015-01-01 02:00:00          Jan            EWR        0    Y  3.0  10.0   

   temp  dewp     slp  pcp 01  pcp 06  pcp 24   sd  
0  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
1  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
2  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
3  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
4  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
5  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
6  30.0   7.0  1023.5     0.0     0.0     0.0  0.0  
7  30.0   6.0  1023.0     0.0     0.0     0.0  0.0  
8  30.0   6.0  1023.0     0.0     0.0     0.0  0.0  
9  30.0   6.0  1023.0     0.0     0.0     0.0  0.0


3) Проверьте, сколько всего строк и столбцов имеется в датасете.

In [ ]:
taxi.shape

(29101, 14)

Строк - 29 101<br>Столбцов - 14

4) Давайте посмотрим на типы колонок. Все ли из них считались правильно?<br> Назовите тип, преобладающий в датасете.

In [ ]:
taxi.dtypes

pickup_dt        object
pickup_month     object
borough          object
pickups           int64
hday             object
spd             float64
vsb             float64
temp            float64
dewp            float64
slp             float64
pcp 01          float64
pcp 06          float64
pcp 24          float64
sd              float64
dtype: object

В датасете преобладает тип float


5) В названиях столбцов pcp 01, pcp 06, pcp 24 встречается пробел. Это не очень удобно, если в дальнейшем вы планируете обращаться к колонкам через точку, без использования кавычек и скобочек. <br>
Замените пробел в названиях на знак нижнего подчеркивания.


In [ ]:
taxi.columns = taxi.columns.str.replace(' ', '_')
taxi.head(5)

pickup_dt pickup_month    borough  pickups hday  spd   vsb  temp  \
0  2015-01-01 01:00:00          Jan      Bronx      152    Y  5.0  10.0  30.0   
1  2015-01-01 01:00:00          Jan   Brooklyn     1519    Y  5.0  10.0  30.0   
2  2015-01-01 01:00:00          Jan        EWR        0    Y  5.0  10.0  30.0   
3  2015-01-01 01:00:00          Jan  Manhattan     5258    Y  5.0  10.0  30.0   
4  2015-01-01 01:00:00          Jan     Queens      405    Y  5.0  10.0  30.0   

   dewp     slp  pcp_01  pcp_06  pcp_24   sd  
0   7.0  1023.5     0.0     0.0     0.0  0.0  
1   7.0  1023.5     0.0     0.0     0.0  0.0  
2   7.0  1023.5     0.0     0.0     0.0  0.0  
3   7.0  1023.5     0.0     0.0     0.0  0.0  
4   7.0  1023.5     0.0     0.0     0.0  0.0

6) Вас попросили узнать, сколько записей (строк) в датафрейме относятся к району Манхэттен (Manhattan). Получить ответ на данный вопрос можно было бы, например, с помощью следующей команды:<br>
taxi.query("borough == 'Manhattan'").shape[0]<br>
А если мы хотим посмотреть, сколько раз встречается каждый из районов? Неужели придется использовать подобную конструкцию для каждого уровня переменной? Здесь на помощь приходит метод value_counts(), подробнее о котором можно почитать в конспекте! <br>
Сколько раз в данных встречается район Бруклин (Brooklyn)?

In [ ]:
taxi['borough'].value_counts()

borough
Bronx            4343
Brooklyn         4343
EWR              4343
Manhattan        4343
Queens           4343
Staten Island    4343
Name: count, dtype: int64

Бруклин встречается 4343 раза

7) Следующая задача: выяснить, из какого района было совершено наибольшее количество поездок за весь период. <br>
Для начала, посчитайте общее количество поездок (pickups), без группировки.

In [ ]:
total_pickups = taxi.pickups.sum()
total_pickups

14265773

8) А теперь сгруппируйте данные по району (borough) и укажите, из какого пункта было совершено больше всего поездок.


In [ ]:
poezdki = taxi.query('pickups > 0').\
    groupby('borough').\
    aggregate({'pickups': 'count'}).\
    sort_values('pickups', ascending=False)
poezdki

pickups
borough               
Brooklyn          4335
Manhattan         4335
Queens            4335
Bronx             4334
Staten Island     3052
EWR                100

9) ⭐️Задание со звёздочкой!⭐️<br>
Еще несколько полезных методов в pandas! Возможно, в предыдущем степе вы просто посмотрели на данные или отсортировали значения. Ускорить данный процесс в дальнейшем помогут методы idxmin() и idxmax(), которые возвращают индекс минимального или максимального значения.<br>
Сохраните название района с наименьшим числом поездок в переменную min_pickups, применив подходящий метод.

In [ ]:
min_pickups = poezdki.idxmin()
min_pickups

pickups    EWR
dtype: object


10) Продолжим изучение данных и посмотрим на число поездок в выходные дни. Сгруппируйте данные по двум признакам: району города и является ли день выходным (колонки borough и hday). Сравните среднее число поездок, и назовите районы, из которых по праздникам в среднем поступает больше заказов, чем в обычные дни.

In [ ]:
new_table = taxi.groupby(['borough', 'hday']).\
                            agg({'pickups': 'mean'})
new_table

pickups
borough       hday             
Bronx         N       50.771073
              Y       48.065868
Brooklyn      N      534.727969
              Y      527.011976
EWR           N        0.023467
              Y        0.041916
Manhattan     N     2401.302921
              Y     2035.928144
Queens        N      308.899904
              Y      320.730539
Staten Island N        1.606082
              Y        1.497006

11) Для каждого района посчитайте число поездок по месяцам. Отсортируйте полученные значения по убыванию и сохраните результирующий датафрейм в pickups_by_mon_bor. Обратите внимание, что итоговый датасет должен состоять из 3-х колонок – pickup_month, borough, pickups.

In [ ]:
pickups_by_mon_bor = taxi.groupby(['borough', 'pickup_month']).\
                                agg({'pickups':'sum'}).\
                                sort_values('pickups', ascending=False)
pickups_by_mon_bor.head(10)

pickups
borough   pickup_month         
Manhattan Jun           1995388
          May           1888800
          Feb           1718571
          Mar           1661261
          Apr           1648278
          Jan           1455543
Brooklyn  Jun            482466
          May            476087
          Apr            378095
          Mar            346726

12) 💀 Сложное задание! 💀<br>
Теперь задача посложнее! Попрактикуемся в написании собственных функций и их применении к датафрейму. Поскольку данные о поездках в Нью-Йорке, температура представлена в градусах Фаренгейта.<br>
Напишите функцию temp_to_celcius, которая получает на вход колонку с температурой в °F, и возвращает значения, переведенные в градусы Цельсия.
Формула:<br>
<br>
Celsius = ((Fahrenheit-32) *5.0)/(9.0)<br>
 ​
Пример работы:<br>
> taxi['temp'][:5]<br>
0  &nbsp; &nbsp;   30.0<br>
1  &nbsp; &nbsp;   30.0<br>
3  &nbsp; &nbsp;    30.0<br>
4  &nbsp; &nbsp;   30.0<br>
5  &nbsp; &nbsp;   30.0<br>

> taxi['temp_C'] = temp_to_celcius(taxi['temp'])<br>
> taxi['temp_C'][:5]<br>
0 &nbsp;  &nbsp;  -1.111111<br>
1 &nbsp; &nbsp;   -1.111111<br>
3 &nbsp; &nbsp;   -1.111111<br>
4  &nbsp; &nbsp;  -1.111111<br>
5  &nbsp; &nbsp;  -1.111111<br>

In [ ]:
taxi['temp'][:5]

0    30.0
1    30.0
2    30.0
3    30.0
4    30.0
Name: temp, dtype: float64

In [ ]:
def temp_to_celcius(Fahrenheit):
    Celsius = ((Fahrenheit-32) *5.0)/(9.0)
    return Celsius

taxi['temp_C'] = temp_to_celcius(taxi['temp'])
taxi['temp_C'][:5]

0   -1.111111
1   -1.111111
2   -1.111111
3   -1.111111
4   -1.111111
Name: temp_C, dtype: float64